cobalt/common/libc/time/tzset.cc

```diff
@@ -0,0 +1,648 @@
+// Copyright 2025 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <limits.h>
+#include <pthread.h>
+#include <string.h>
+#include <time.h>
+
+#include <array>
+#include <charconv>
+#include <chrono>
+#include <ctime>
+#include <memory>
+#include <optional>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+#include "starboard/common/log.h"
+#include "starboard/log.h"
+#include "unicode/basictz.h"
+#include "unicode/calendar.h"
+#include "unicode/gregocal.h"
+#include "unicode/putil.h"
+#include "unicode/simpletz.h"
+#include "unicode/timezone.h"
+#include "unicode/tznames.h"
+#include "unicode/tztrans.h"
+#include "unicode/ucal.h"
+#include "unicode/udata.h"
+#include "unicode/unistr.h"
+#include "unicode/utypes.h"
+
+#include "cobalt/common/libc/no_destructor.h"
+#include "cobalt/common/libc/tz/parse_posix_tz.h"
+#include "starboard/time_zone.h"
+
+namespace cobalt {
+namespace common {
+namespace libc {
+namespace time {
+
+#ifndef TZNAME_MAX
+#define TZNAME_MAX 6
+#endif
+
+namespace {
+// A thread-local flag to detect re-entrant calls to tzset().
+thread_local bool g_in_tzset = false;
+
+// A scoped guard to ensure the g_in_tzset flag is reset on exit.
+class TzsetGuard {
+ public:
+  TzsetGuard() {
+    SB_CHECK(!g_in_tzset) << "tzset() re-entrancy detected.";
+    g_in_tzset = true;
+  }
+  ~TzsetGuard() { g_in_tzset = false; }
+};
+}  // namespace
+
+// A class to encapsulate timezone-related logic and state.
+class Timezone {
+ public:
+  static void TZSet();
```

Comment:
nit: let's be consistent about TZset/Tzset capitalization (i.e. there's TzsetGuard above)